### PR TITLE
Smooth lighting for (virtually) everything

### PR DESCRIPTION
**Pull request by numberZero**
See #104 
This patch adds smooth lighting to mesh and nodebox nodes. There may be lighting bugs (esp. on mapblock boundaries), but crashes are fixed (I hope so...), and such lighting should be way better than current smooth-sharp mix anyway.

### CODE REVIEW

File: src/content_mapblock.cpp [hunk 1]
```diff
@@ -30,22 +30,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 #include "noise.h"
 
+#define SMOOTH_LIGHTING_OVERSIZE 0.0f
+
+typedef u16 LightFrame[8];
+
+static const v3s16 light_dirs[8] = {
```
Comment (sfan5) [inline src/content_mapblock.cpp]: what's this for? add a comment please

File: src/content_mapblock.cpp [hunk 2]
```diff
@@ -30,22 +30,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 #include "noise.h"
 
+#define SMOOTH_LIGHTING_OVERSIZE 0.0f
```
Comment (sfan5) [inline src/content_mapblock.cpp]: what's this for? add a comment please

Comment (sfan5) [inline src/content_mapblock.cpp]: what's this for? add a comment please

File: src/content_mapblock.cpp (hunk 1 above)
Comment (sfan5) [inline src/content_mapblock.cpp]: what's this for? add a comment please

File: src/content_mapblock.cpp [hunk 3]
```diff
@@ -30,22 +30,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 //              for the opposite corners of each face - therefore, there
 //              should be (2+2)*6=24 values in the list. Alternatively, pass
 //              NULL to use the entire texture for each face. The order of
 //              the faces in the list is up-down-right-left-back-front
 //              (compatible with ContentFeatures). If you specified 0,0,1,1
 //              for each face, that would be the same as passing NULL.
-void makeCuboid(MeshCollector *collector, const aabb3f &box,
-	TileSpec *tiles, int tilecount, video::SColor &c, const f32* txc)
+static void makeSmoothLightedCuboid(MeshCollector *collector, const aabb3f &box,
+	TileSpec *tiles, int tilecount, video::SColor *c, const f32* txc)
```
Comment (sfan5) [inline src/content_mapblock.cpp]: should be `const video::Scolor *c`, no?

File: src/content_mapblock.cpp [hunk 4]
```diff
@@ -164,6 +178,52 @@ void makeCuboid(MeshCollector *collector, const aabb3f &box,
...
+}
+
+// Gets the base lighting values for a node
+//  p      - node position (absolute)
+//  data   - ...
+//  lights - resulting (opaque) data
+static void getSmoothLightFrame(LightFrame &lights, const v3s16 &p, MeshMakeData *data)
+{
+		for (int k = 0; k != 8; ++k)
+			lights[k] = getSmoothLight(p, light_dirs[k], data);
```
Comment (sfan5) [inline src/content_mapblock.cpp]: indented too much

File: src/content_mapblock.cpp [hunk 5]
```diff
@@ -164,6 +178,52 @@ void makeCuboid(MeshCollector *collector, const aabb3f &box,
...
+//  light_source - node's light source property
+static video::SColor blendLight(const LightFrame &lights, const core::vector3df& vertex_pos, u8 light_source)
+{
+	f32 x = core::clamp(vertex_pos.X, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 y = core::clamp(vertex_pos.Y, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 z = core::clamp(vertex_pos.Z, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 lightA = 0.0;
+	f32 lightB = 0.0;
+	for (int k = 0; k != 8; ++k)
+	{
```
Comment (sfan5) [inline src/content_mapblock.cpp]: brace should be on previous line

File: src/content_mapblock.cpp [hunk 6]
```diff
@@ -164,6 +178,52 @@ void makeCuboid(MeshCollector *collector, const aabb3f &box,
...
+//  vertex_pos   - vertex position in the node (coordinates are clamped to [0.0, 1.0] or so)
+//  light_source - node's light source property
+static video::SColor blendLight(const LightFrame &lights, const core::vector3df& vertex_pos, u8 light_source)
+{
+	f32 x = core::clamp(vertex_pos.X, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 y = core::clamp(vertex_pos.Y, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 z = core::clamp(vertex_pos.Z, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 lightA = 0.0;
+	f32 lightB = 0.0;
+	for (int k = 0; k != 8; ++k)
```
Comment (sfan5) [inline src/content_mapblock.cpp]: use `k < 8` please (compiler will compile it to the same thing)

File: src/content_mapblock.cpp [hunk 7]
```diff
@@ -1655,8 +1715,17 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
...
-			video::SColor c = MapBlock_LightColor(255, l, f.light_source);
+			LightFrame lights;
+			video::SColor colors[8];
+
+			if(data->m_smooth_lighting) {
+				getSmoothLightFrame(lights, blockpos_nodes + p, data);
+			} else {
+				u16 l = getInteriorLight(n, 1, nodedef);
+				video::SColor c = MapBlock_LightColor(255, l, f.light_source);
+				for (int j = 0; j != 8; ++j)
```
Comment (sfan5) [inline src/content_mapblock.cpp]: use `j < 8` please

File: src/mapblock_mesh.cpp [hunk 8]
```diff
@@ -242,7 +242,7 @@ static u16 getSmoothLightCombined(v3s16 p, MeshMakeData *data)
 
 	for (u32 i = 0; i < 8; i++)
 	{
-		const MapNode &n = data->m_vmanip.getNodeRefUnsafeCheckFlags(p - dirs8[i]);
+		MapNode n = data->m_vmanip.getNodeNoExNoEmerge(p - dirs8[i]);
```
Comment (sfan5) [inline src/mapblock_mesh.cpp]: why is the `const` removed here?

File: src/content_mapblock.cpp [hunk 9]
```diff
@@ -1747,7 +1809,15 @@ void mapblock_mesh_generate_special(MeshMakeData *data,
 					// front
 					tx1, 1-ty2, tx2, 1-ty1,
 				};
-				makeCuboid(&collector, box, tiles, 6, c, txc);
+				if (data->m_smooth_lighting) {
+					for (int j = 0; j != 8; ++j) {
+						f32 x = j & 4 ? dx2 : dx1;
```
Comment (rubenwardy) [inline src/content_mapblock.cpp]: same here

File: src/content_mapblock.cpp [hunk 10]
```diff
@@ -30,22 +30,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 #include "noise.h"
 
+#define SMOOTH_LIGHTING_OVERSIZE 0.0f
+
+typedef u16 LightFrame[8];
```
Comment (rubenwardy) [inline src/content_mapblock.cpp]: I don't like how you're hiding an array in a typedef, but meh :P

File: src/content_mapblock.cpp [hunk 11]
```diff
@@ -164,6 +178,52 @@ void makeCuboid(MeshCollector *collector, const aabb3f &box,
...
+	f32 x = core::clamp(vertex_pos.X, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 y = core::clamp(vertex_pos.Y, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 z = core::clamp(vertex_pos.Z, 0.0f - SMOOTH_LIGHTING_OVERSIZE, 1.0f + SMOOTH_LIGHTING_OVERSIZE);
+	f32 lightA = 0.0;
+	f32 lightB = 0.0;
+	for (int k = 0; k != 8; ++k)
+	{
+		u32 light1A = lights[k] & 0xff;
+		u32 light1B = lights[k] >> 8;
+		f32 dx = k & 4 ? x : 1 - x;
```
Comment (rubenwardy) [inline src/content_mapblock.cpp]: please use parenthesis to make code more readable, ie:

```cpp
(k & 4) ? x : 1 - x
```

File: src/mapblock_mesh.cpp (hunk 8 above)
Comment (numberZero) [inline src/mapblock_mesh.cpp]: Because unlike `getNodeRefUnsafeCheckFlags`, `getNodeNoExNoEmerge` returns an object and not a reference. And, well, const-reference to a 4-byte object looks weird anyway.

File: src/content_mapblock.cpp (hunk 10 above)
Comment (numberZero) [inline src/content_mapblock.cpp]: It is intended to be something opaque, used only to transfer some data from `getSmoothLightFrame` to `blendLight`. E.g. it could contain the node’s own light level, or its `light_source` value (it would be a structure then), etc.
Surely I can change that, but only is you consider that necessary.

**Comment by Fixer-007** [comment]
Looks very-very nice in testing.

File: src/content_mapblock.cpp [hunk 12]
```diff
@@ -30,22 +30,36 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
 //  txc       - texture coordinates - this is a list of texture coordinates
 //              for the opposite corners of each face - therefore, there
 //              should be (2+2)*6=24 values in the list. Alternatively, pass
 //              NULL to use the entire texture for each face. The order of
 //              the faces in the list is up-down-right-left-back-front
 //              (compatible with ContentFeatures). If you specified 0,0,1,1
 //              for each face, that would be the same as passing NULL.
-void makeCuboid(MeshCollector *collector, const aabb3f &box,
-	TileSpec *tiles, int tilecount, video::SColor &c, const f32* txc)
+static void makeSmoothLightedCuboid(MeshCollector *collector, const aabb3f &box,
```
Comment (numberZero) [inline src/content_mapblock.cpp]: According to my dictionary, both **lit** and **lighted** are correct)

File: src/content_mapblock.cpp (hunk 2 above)
Comment (numberZero) [inline src/content_mapblock.cpp]: This is intended for oversized nodes. It’s the distance the light value being extrapolated linearly, being constant at greater distance from the (logical) node boundary.
Not sure how to make a good comment from that, though.

File: src/content_mapblock.cpp (hunk 10 above)
Comment (numberZero) [inline src/content_mapblock.cpp]: UPD: converted `LightFrame` to a structure. That would be better IMO.

**Comment by numberZero** [comment]
@Fixer-007 
> lines on stairs

Looks strange. I though that’s due to some non-linearity in other light processing, but it seems there is some another reason.

> homedecor:projection_screen is not smoothed

It works for meshes and nodeboxes only yet, and `homedecor:projection_screen` is neither, it’s `signlike`. Didn’t thought these worth smoothing too.

> flowing liquids still not smoothed

Uh oh, the same reason here, but it certainly needs to be smoothed.

**Comment by VanessaE** [comment]
If I remember right, I also used signlike for home decor's "ceiling paint" and "ceiling tiles" nodes.

**Comment by numberZero** [comment]
@Fixer-007 It was a miracle it worked at all for meshes, the code was totally broken. I didn’t mention the `BS` constant... (now fixed)
@sfan5 @rubenwardy updated

So should I add support of other node types in this PR? And is smooth lighting desirable for all node types? It can easily be added the same way I did for nodeboxes, but that would bloat the code considerably.

**Comment by paramat** [comment]
How does this work? Does it calculate the face as if it was a cube then project that onto the nodebox face that actually exists?

File: src/content_mapblock.cpp (hunk 12 above)
Comment (paramat) [inline src/content_mapblock.cpp]: In this context lighted seems ok.

File: src/content_mapblock.cpp [hunk 13]
```diff
@@ -30,22 +30,43 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+	TileSpec *tiles, int tilecount, const video::SColor c[8], const f32* txc)
```
Comment (sfan5) [inline src/content_mapblock.cpp]: what @rubenwardy said is not actually true, adding the array size to the call **does not** protect from wrongly-sized array.
This code:
```c++
#include <iostream>

void thing(const int array[8]) {
	std::cout << "first element: " << array[0] << std::endl;
}

int main() {
	int a1[8] = {0,0,0,0,0,0,0,0};
	thing(a1);

	int a2[2] = {1,1};
	thing(a2); //too small

	int a3[16] = {2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2};
	thing(a3); //too big

	int a4 = 3;
	thing(&a4); // not an array
}
```
...compiles without **any** warnings using `-Wall -Wextra`

tl;dr change it back to `const video::SColor *c` to avoid confusion

File: src/content_mapblock.cpp (hunk 12 above)
Comment (ahkok) [inline src/content_mapblock.cpp]: Both are correct, however 'lit' sounds a lot better to me.

`the sun lighted up the sky` ....

**Comment by sofar** [comment]
@numberZero 
I'd like to note that non-smooth lighting is highly apparent for nodes that connect fully to adjacent nodes. This is why nodebox (slabs) nodes, mesh (stairs) are nodes that show non-smooth light the most - they're often used in almost all builds.

From that perspective, it's easy to see that rails (which is `raillike`) would also be showing the problem as well, especially since people will build long and connected rail tracks.

Interestingly signs are not signlike (they're nodebox already).

Liquids and glasses are already smoothly lit, aren't they?

Only thing I can think of is `allfaces_optional` - used everywhere by leaves.

**Comment by numberZero** [comment]
@paramat it calculates lighting at (logical) node vertices the same way old code does for solid nodes (except that it uses all 8 vertices, and some more checks are necessary now, so it uses `getNodeNoExNoEmerge` now), then interpolates trilinearly for each real vertex. For oversized nodes (e.g. mesh doors in MTG) it can extrapolate too, but that may or may not look well, probably that should be configurable.

I know 2 problems with that:
1. It is heavy. Lighting calculations were never easy, but this code requires considerably more processing, especially for complex meshes.
2. There are small lighting bugs. That’s unavoidable with current mesh processing system: to calculate lighting for a node properly, it must know light levels at all the adjacent nodes. That may not be possible on the mapblock boundary.

The solution could be to use a 3D texture as lighting map and a shader to apply it. That would speed up the mesh generation considerably (as no interpolation would be necessary there, and the cache could be used to its full potential), and thus the 2nd problem could be solved in a straightforward way: the texture could be just recreated on load of an adjacent mapblock, as that would be very fast. But there is one problem: Irrlicht doesn’t seem to support 3D textures at all.

File: src/content_mapblock.cpp (hunk 13 above)
Comment (numberZero) [inline src/content_mapblock.cpp]: @sfan5 Well, it doesn’t, but it does show that an 8-element array must be passed.

**Comment by juhdanad** [comment]
@numberZero a simple 3D texture is not enough. Imagine three stone blocks at `(1,0,0)`, `(0,1,0)` and `(0,0,1)`. Then the common corner of `(0,0,0)` and `(1,1,1)` should have two light levels, since the two blocks are completely separated. (note: Minetest's lighting is a bit buggy, as currently the corners always have one common light level)

**Comment by numberZero** [comment]
@juhdanad I don’t mean “photorealistic lighting,” I just described a way to make current smooth lighting work for everything at once. The bug you described is definitely not the worst bug MT lighting model has. But such bugs are unavoidable in a voxel game.

**Comment by Fixer-007** [comment]
On my core i3 + amd/ati hd6870 I checked fps with and without this PR in two scenes with lots of slabs and stairs and it was the same honestly, maybe other will have different results.

**Comment by numberZero** [comment]
Merge is easier than rebase. I hope this will not mess the history up too much. Now compiling and testing.
@juhdanad I kept your new shaded cuboid for non-smooth lighting, so there are 2 cuboid styles now =/

**Comment by numberZero** [comment]
@juhdanad added me some work...

**Comment by Fixer-007** [comment]
Compiler error:
```
[ 48%] Building CXX object src/CMakeFiles/minetest.dir/content_sao.cpp.obj
/media/sf_Exchange/minetest/build/minetest/src/content_mapblock.cpp: In function ‘irr::video::SColor blendLight(const LightFrame&, const vector3df&)’:
/media/sf_Exchange/minetest/build/minetest/src/content_mapblock.cpp:392:59: error: ‘MapBlock_LightColor’ was not declared in this scope
  return MapBlock_LightColor(255, light, frame.light_source);
                                                           ^
/media/sf_Exchange/minetest/build/minetest/src/content_mapblock.cpp:393:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
```

**Comment by numberZero** [comment]
@Fixer-007 I know, fixing now.

**Comment by numberZero** [comment]
@Fixer-007 @sfan5 Rebased, tested, works.

**Comment by Fixer-007** [comment]
Compiled, having problem with signs now, they look like this https://i.imgur.com/TUGzBto.png

**Comment by numberZero** [comment]
@Fixer-007 Fixed signs (they appear to be the only nodes using the nodebox drawtype).
But with the latest changes solid nodes have different lighting, so I have to change mesh processing now (stairs and slabs use the mesh drawtype)

**Comment by Fixer-007** [comment]
Will try again now :)

**Comment by Fixer-007** [comment]
* Slabs seems to have problems again

**Comment by numberZero** [comment]
@Fixer-007 check again, please. It should look perfectly now.

**Comment by Fixer-007** [comment]
OK, give me 5 minutes.

**Comment by numberZero** [comment]
(I don’t mean liquids, `signlike`, etc., I hadn’t enough time to fix those too, being adopting juhdanad’s changes)

**Comment by Fixer-007** [comment]
Yeah, slabs and stairs are fixed now

**Comment by numberZero** [comment]
@juhdanad You may need to revise the `applyFacesShading` function, it should give reasonable values for any (unit) vector as it is used for meshes too now.

**Comment by sofar** [comment]
can't seem to apply to my branch:

```
 git am 5098.patch
Applying: Smooth lighting for nodeboxes!
error: patch failed: src/content_mapblock.cpp:164
error: src/content_mapblock.cpp: patch does not apply
error: patch failed: src/nodedef.cpp:728
error: src/nodedef.cpp: patch does not apply
Patch failed at 0001 Smooth lighting for nodeboxes!
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

**Comment by sofar** [comment]
- All my torches are upside down. (mtg's 3d torches)
- With shaders enabled, my entire world is daylight bright at midnight.

This is with ccf670214f4982d22254d47237ae8beb74dd8867 (a clone of your branch)

**Comment by paramat** [comment]
>  then interpolates trilinearly for each real vertex.

Ok thanks, so the processing depends on the number of vertices in the nodebox / mesh. I expect that is not a problem for slabs and stairs since they only have a few more vertices. Nodeboxes / meshes in default MTG are always simple so that won't be a problem in MTG.
Something to consider: should there be a bool to disable smooth lighting for complex meshes and/or those not badly affected by non-smooth lighting? I'm not requesting this and am unsure, just something to consider.

This work is very much appreciated.

**Comment by paramat** [comment]
The stress test for performance would be at spawn in VanessaE's creative surver, can anyone test performance impact?

File: src/content_mapblock.cpp [hunk 14]
```diff
@@ -30,6 +30,27 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "log.h"
 #include "noise.h"
 
+// Distance of light extrapolation (for oversized nodes)
+// After this distance, it gives up and considers light level constant
+#define SMOOTH_LIGHTING_OVERSIZE 1.0
```
Comment (paramat) [inline src/content_mapblock.cpp]: I'm wondering if oversize extrapolation is worth doing, since it will probably not match lighting anyway and not doing this may look just as good. Not many nodes are oversized.

**Comment by VanessaE** [comment]
To clarify, @paramat refers to the "old" spawn area.  Get there via the "N/S/W/E crossroads" (or something like that) destination in the "Gridknot" Travelnet network.

**Comment by sofar** [comment]
It seems that the bug I'm seeing with shaders enabled is that the light bank's day value is always used entirely. In caves, the areas are properly dark. The problem is only on the surface.

Is nobody else seeing the torch issues?

**Comment by sofar** [comment]
I'm seeing the upside torches on VE dreambuilder as well, not just on my own servers.

**Comment by numberZero** [comment]
@sofar 
> can't seem to apply to my branch

The `master` branch was updated changed yesterday, I had to rebase my patch. See the `dev-working-smooth-lighting-1` tag, it should apply cleanly on older version.

> All my torches are upside down. (mtg's 3d torches)
> With shaders enabled, my entire world is daylight bright at midnight.

Strange. Everything looks fine on my machine (in singleplayer mode).
Do you use clean `ccf6702`? It seems to use different format for light level coding (probably due to #4986).

File: src/content_mapblock.cpp (hunk 14 above)
Comment (numberZero) [inline src/content_mapblock.cpp]: I’m not sure too, but that is very simple, so I decided to leave such option.

**Comment by sofar** [comment]
```
$ git describe
0.4.14-516-gccf670214
```

so a clean checkout. I just did a full `make clean` and rebuild, and the light issue appears solved, sorry.

Torches still upside down, though. No `torches` mod loaded. freakish.

**Comment by numberZero** [comment]
@sofar The reason is that latest shaders (updated yesterday) have different input format.
This is what I got with my patch but shaders one day old:
https://imgur.com/a/EaZNu
And this is with the latest shaders (from the `master` branch), with different options:
https://imgur.com/a/FU3Ex

Investigating what could cause the torches bug. What settings do you use?

**Comment by numberZero** [comment]
@sofar Found it. It appears when the mesh cache is enabled.

**Comment by sofar** [comment]
oh, yeah, that would explain that.

**Comment by sofar** [comment]
Yup, the cache indeed was the issue.

**Comment by sofar** [comment]
Well, gee, this looks good then!

**Comment by VanessaE** [comment]
Mesh cache should be removed entirely imho.  it costs huge amounts of RAM if you have lots of mods, with little benefit.

**Comment by numberZero** [comment]
@VanessaE That’s out of scope of this PR IMO

**Comment by VanessaE** [comment]
I know.  Just bears mentioning.

**Comment by numberZero** [comment]
@Fixer-007 @sofar Updated again, now should work for everything: liquids, glass, leaves, etc.

It probably needs some refactoring, but works as is.

**Comment by juhdanad** [comment]
This works, great job!

However, a refactor is really needed, as there are members and variables as `lightsA`, `lightsB`, `txc` and `dx1` that are not documented. Also there are more functions to draw a cuboid, these should be generalized.

**Comment by numberZero** [comment]
Yeah, there are probably too many cuboid-making functions) But the problem is that the proper refactoring is not straightforward, it would require significant code change. Should that go in this PR? @paramat @sfan5 @rubenwardy your opinion?

**Comment by paramat** [comment]
I would prefer a separate PR for that, to make this one easier to review.

**Comment by paramat** [comment]
Undocumented or unclear variable names are not an urgent problem, consider that those who know the code are used to these variables and know what they are. A refactor is fairly low priority, best delay it a while to reduce the amount of reviewing needed.

**Comment by Fixer-007** [comment]
Entities (like carts) are not smoothed, right?

**Comment by Fixer-007** [comment]
Some comparisons I did, very impressive results:
https://i.imgur.com/ieFbtQk.jpg vs https://i.imgur.com/aqKgt2n.jpg
https://i.imgur.com/P3ABT1G.png vs https://i.imgur.com/g88u4my.png
https://i.imgur.com/ZpzJbHG.jpg vs https://i.imgur.com/obdZq0G.jpg
https://i.imgur.com/jyviyEi.png vs https://i.imgur.com/aQuF374.png

**Comment by MarkuBu** [comment]
Great job. Can't wait to see this in the dev version

**Comment by numberZero** [comment]
@Fixer-007 Entities can’t be smoothed this way, that’s totally impossible. A 3D texture could help, though, but Irrlicht doesn’t support that AFAIK.

**Comment by nerzhul** [comment]
Very nice work.
Just a question, is it possible to make those function be owned by a class ?

**Comment by numberZero** [comment]
@nerzhul Surely that’s possible, but that would be pointless without normal refactoring which should be done in a separate PR.
As I imagine, there should be 3 classes: `MeshBuilder`, `SharpLighter` and `SmoothLighter`; the `mapblock_mesh_generate_special` function (it’s 1.5k lines long...) should be split into methods of `MeshBuilder`; the various cuboid-making functions would be refactored too, spreading the code across these classes in a more logical way. But that requires a lot of time and effort; nevertheless, I can try to do that (after a few days) as a separate PR.

**Comment by sofar** [comment]
I'm with paramat and think that refactoring on top of this patch would get way too large.

I'd like to merge as-is. Even if not all the node types are not included yet (I hadn't thought of the old fence type since it's not used anymore in `minetest_game` even).

It feels like an entire new game to me.

:+1:

Further requested changes for future PRs:

- smooth lighted entities (if possible mesh, cuboid and sprites, they're all used)
- **smooth lighted particles**
- refactoring

**Comment by numberZero** [comment]
@sofar It should support all the node types available, including fences, liquids, signs, etc.
I can’t do entity shading, the only way I know requires a feature not present in Irrlicht.

**Comment by sofar** [comment]
yeah, something for later for sure. @nerzhul if you're OK with this, please merge or I will.

**Comment by nerzhul** [comment]
I'm looking TV atm please merge @sofar

**Comment by sofar** [comment]
@numberZero can you please do a `git fetch origin; git pull --rebase origin master` and resolve the conflicts in `content_mapblock.cpp` and `nodedef.cpp`, and then re-push to the branch?

You have 15 commits interweaved with several commits, and normally I attempt to rebase myself, but this is too many large chunks that need to be checked.

I'd love to merge but given the size I think we should take care merging this :)

I can squash after, you don't have to do that.

**Comment by numberZero** [comment]
@sofar there are problems with rebase; can I just squash it to a single commit?

**Comment by sofar** [comment]
Only doing a squash in your local branch will not resolve the conflict. Join me in IRC and I'll help you do the rebase.

**Comment by sofar** [comment]
@juhdanad this would best be rebased by both of you together, can you help @numberZero ?

**Comment by sofar** [comment]
33e0eedbfb116111fa79cbc506d9e94ffbb1543b

**Comment by sofar** [comment]
I know there's a few bugreports out that we can close with this now. If you spot one, please let us know so we can close them! thanks!

**Comment by Fixer-007** [comment]
Only minor and rare problem for me remains this: https://i.imgur.com/00WhN42.png But it is insanely hard to spot IRL builds (I spotted none).

**Comment by numberZero** [comment]
@Fixer-007 is that the correct screenshot for the problem you mean? I know there may be a similar problem at mapblock boundary, but that has another reason and is virtually unavoidable.

**Comment by Fixer-007** [comment]
Here is the world that has this problem, I'm pointing at stairs with this line: 
[PR5098.zip](https://github.com/minetest/minetest/files/728159/PR5098.zip)
